[PyTorch] Speed up decomposed quantize_per_channel

swolchok opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #133029

Similar to D60871396 (#132828).

Differential Revision: [D60978385](https://our.internmc.facebook.com/intern/diff/D60978385/)

swolchok committed:
[PyTorch] Speed up decomposed quantize_per_channel

Similar to D60871396 (#132828).

Differential Revision: [D60978385](https://our.internmc.facebook.com/intern/diff/D60978385/)

[ghstack-poisoned]

vadimkantorov commented on torch/ao/quantization/fx/_decomposed.py:
```diff
@@ -589,15 +589,13 @@ def quantize_per_channel(
+    scales = scales.view(new_shape)
+    zero_points = zero_points.view(new_shape)
 
+    res = torch.clamp(torch.round(input * (1.0 / scales)) + zero_points, quant_min, quant_max)
```
does it not change how zero_points / scales need to be broadcasted?

e.g. for this new version to work, `permute` should happen to the last axis instead of `axis_zero`, rigth?

for this code to work as is, broadcast needs to happen "on the right" (which is not supported), no?

swolchok replied:
I don't have a great understanding of PyTorch broadcasting, but as evidence that this works, I submit 1) the tests, which haven't failed yet, and 2) https://github.com/pytorch/pytorch/pull/132828 , which I linked in the summary.

vadimkantorov replied:
Oh, I now see in your code you're manually adding singular dims to `scale`/`zero_point`. I think a simpler way would be to move the feature-dims to the last dim (instead of zero'th dim), and then there is no need for inserting singular dims...

swolchok replied:
> I think a simpler way would be to move the feature-dims to the last dim (instead of zero'th dim), and then there is no need for inserting singular dims as broadcasting would handle missing singular dims on the fly...

I don't understand how to do this. Please feel free to send a follow-up; it sounds easy!

vadimkantorov replied:
I think, this would work:
`res = torch.clamp(torch.round(input.transpose(axis, -1) * (1.0 / scales)) + zero_points, quant_min, quant_max).transpose(axis, -1)`

Also, should be noted, that both my solution and your solution change the contiguity pattern (my solution makes it `channels_last` and your solution moves the feature dim to front and makes it contig). @jerryzh168, am I correct and does it matter in this context? (or maybe `*` and `+` would propagate left-arg strides?)

swolchok committed:
lint on "[PyTorch] Speed up decomposed quantize_per_channel"


Similar to D60871396 (#132828).

Differential Revision: [D60978385](https://our.internmc.facebook.com/intern/diff/D60978385/)

[ghstack-poisoned]

cccclai reviewed (approved):
Looks good. @jerryzh168 any specific reason `zero_point` is not optional?

jerryzh168 commented:
> Looks good. @jerryzh168 any specific reason `zero_point` is not optional?

just historically we don't have observers that has zero_point == None, but this is supported in the new quant primitive ops in torchao

swolchok commented:
@pytorchbot merge
